test/test.dart

```diff
@@ -84,157 +86,192 @@ const kScreenWidth = 450.0;
-  ///  5. un-pumps the screen
-  Future<void> runAppTest(AsyncCallback callback, {AsyncCallback? goldenCaptureCallback}) async {
+  /// Run a test that uses the App UI using the following procedure:
+  ///  1. Initializes the application, optionally running [initialization] when all fakes are set up
+  ///     before the [ContentControl] is initialized.
+  ///  2. Optionally, runs [postInitialization] after [ContentControl] is initialized.
+  ///  3. Pumps the app.
+  ///  4. Runs the test from the [callback].
+  ///  5. Optionally, runs [goldenCaptureCallback].
+  ///  6. Stops and disposes the player.
+  ///  7. Un-pumps the screen and flushes all micro-tasks and stream events.
+  Future<void> runAppTest(
+    AsyncCallback callback, {
+    AsyncCallback? goldenCaptureCallback,
+    VoidCallback? initialization,
+    VoidCallback? postInitialization,
+  }) async {
     await withClock(binding.clock, () async {
-      // App only supports vertical orientation, so switch tests to use it.
-      await binding.setSurfaceSize(kScreenSize);
-      await pumpWidget(ProviderScope(
-        overrides: [
-          playerInterfaceColorStyleArtColorBuilderProvider
-              .overrideWithValue(FakePlayerInterfaceColorStyleArtColorBuilder())
-        ],
-        child: const App(debugShowCheckedModeBanner: false),
-      ));
-      await pump();
-      await callback();
-      await runAsync(() async {
+      await binding._setUpAppTest(initialization);
+      try {
+        postInitialization?.call();
+        // App only supports vertical orientation, so switch tests to use it.
+        await binding.setSurfaceSize(kScreenSize);
+        await pumpWidget(ProviderScope(
+          overrides: [
+            playerInterfaceColorStyleArtColorBuilderProvider
+                .overrideWithValue(FakePlayerInterfaceColorStyleArtColorBuilder())
+          ],
+          child: const App(debugShowCheckedModeBanner: false),
+        ));
+        await pump();
+        await callback();
+        await goldenCaptureCallback?.call();
+      } finally {
         // Don't leak player state between tests.
-        // Delay needed for proper disposal in some tests.
-        await Future.delayed(const Duration(milliseconds: 1));
         await MusicPlayer.instance.stop();
         await MusicPlayer.instance.dispose();
-      });
-      await goldenCaptureCallback?.call();
-      // Un-pump, in case we have any real animations running,
-      // so the pumpAndSettle on the next line doesn't hang on.
-      await pumpWidget(const SizedBox());
-      // Wait for ui animations.
-      await pumpAndSettle();
+        // Un-pump, in case we have any real animations running,
+        // so the pumpAndSettle on the next line doesn't hang on.
+        await pumpWidget(const SizedBox());
+        // Wait for any asynchronous events and stream callbacks to finish.
+        await pump(const Duration(seconds: 1));
+        // Wait for ui animations.
```

Comment:
I imagine this is referring to "system UI" (but don't remember for sure)

Let's put that into the comment tho

Otherwise pumping out the UI above would already stop animations and we wouldn't need pumpAndSettle